### ISSUE TITLE
Training pipeline of PoseNet

### ISSUE THREAD

**Issue by Aku02** [facebookresearch/banmo#29]
Hello there!

Thanks a lot for sharing your work!

I have a couple of questions:

1.What is the dataset you used to train the PoseNet for root pose initialization? 
2.What is the `occ` from the optical flow model? It seems that it's loaded in dataloader but is not used anywhere in the training.

**Comment by gengshan-y** [comment]
Hi, the details about PoseNet training are in the [paper supplement B.1](https://banmo-www.github.io/banmo-cvpr.pdf).

The training data are rendered on the fly. For more details, see [here](https://github.com/facebookresearch/banmo/blob/1d022945be987a601a36a25c27e71a8c06008bef/nnutils/banmo.py#L755-L791) the training code of PoseNet, as well as the [rendering + augmentation pipeline](https://github.com/facebookresearch/banmo/blob/1d022945be987a601a36a25c27e71a8c06008bef/nnutils/banmo.py#L872-L963). Unfortunately, the sheep mesh is under commercial license and we are not allowed to release it.

`occ` is the flow confidence computed by [forward-backward check](https://github.com/facebookresearch/banmo/blob/1d022945be987a601a36a25c27e71a8c06008bef/dataloader/vidbase.py#L234-L254). It is renamed as `cfd_at_samp` and used [here](https://github.com/facebookresearch/banmo/blob/1d022945be987a601a36a25c27e71a8c06008bef/nnutils/rendering.py#L533) to weight flow reconstruction loss.

**Comment by gengshan-y** [comment]
1. The sheep/SMPL mesh/surface features are from [densepose-CSE](https://github.com/facebookresearch/detectron2/blob/main/projects/DensePose/doc/DENSEPOSE_CSE.md). You may get the vertex features but I'm afraid it's not easy to get their vertex locations if you are outside the organization (even I don't have access to them now). The SMPL mesh is a subdivided version of the original one as noted [here](https://github.com/facebookresearch/detectron2/issues/3233#issuecomment-879145446).

2. Yes.

**Comment by vhvkhoa** [comment]
Hi @gengshan-y, after looking through the training code of PoseNet, I would like to ask that did you only use a single human mesh to train PoseNet for humans, and another sheep mesh to train PoseNet for quadruped animals ? 

If we put aside that there are various types of humans and animals, that might be handled well by the pretrained CSE features. How can your PoseNet predict the root pose very well when the objects make various poses ?

**Comment by gengshan-y** [comment]
Yes.

The initial poses passed into banmo are noisy indeed due to deformations/shape variations. BANMo is updating the root poses during optimization. See [here](https://banmo-www.github.io/aba.html).

**Comment by vhvkhoa** [comment]
Thank you for your quick response. But I would like to clarify a little bit that my question is about the function [forward_warmup](https://github.com/facebookresearch/banmo/blob/1d022945be987a601a36a25c27e71a8c06008bef/nnutils/banmo.py#L755-L791) that you used to pre-train PoseNet, which is stored in human.pth and and quad.pth.
I was wondering if they were trained using a single mesh, e.g., the mesh of resting pose sheep (illustrated in Fig.12 of your paper), with random camera poses generated around it ?

**Comment by gengshan-y** [comment]
That is correct, we use rest shape of the sheep/human only to train PoseNet.

**Comment by sidsunny** [comment]
Hi @gengshan-y . Thanks for your amazing work and for sharing the code! I see that the sheep mesh cannot be released. Can you suggest a way to train PoseNet on our custom mesh? I suppose we cannot get CSE embeddings for our custom mesh.

**Comment by gengshan-y** [comment]
Hi, I think there is no off-the-shelf solution.

The most straightforward way to get vertex embedding is to follow the [CSE](https://arxiv.org/abs/2011.12438) solution, where one can label corresponding 3D keypoints on a canonical mesh, and label 2D keypoints on images, from which you can learn a vertex embedding that corresponds to pixel features.

Starting from a mesh with vertex features, you would be able to train banmo's posenet following the paper.
